Changelog for workers-rs panic recovery (#25285)
cloudflare/cloudflare-docs: src/content/changelog/workers/2025-09-19-workers-rs-panic-recovery.mdx

This ensures more reliable deployents. Automatic panic recovery feature is enabled for all new workers-rs deployments as of version 0.6.5, with no further configuration required.

## Fixing Rust Panics with Wasm Bindgen

Rust Workers are built with Wasm Bindgen, which treats panics as non-recoverable - the entire Wasm application is considered to be in an invalid state.

We now attach a default panic handler in Rust:

```rust
std::panic::set_hook(Box::new(move |panic_info| {
  hook_impl(panic_info);
}));
```

Which is registered by default in the JS initialization:

```js
import { setPanicHook } from "./index.js";
setPanicHook(function (err) {
	console.error("Panic handler!", err);
});
```

When a panic occurs, we then reset the Wasm state to reinitialize the Wasm application back to as it was when the application first started.

## Resetting VM State in Wasm Bindgen

We worked upstream on the Wasm Bindgen project to implement a new [`--experimental-reset-state-function` compilation option](https://github.com/wasm-bindgen/wasm-bindgen/pull/4644) which outputs a new `__wbg_reset_state` function.

> This function clears all internal state related to the Wasm VM, and also ensures object references are uniquely associated with the Wasm instance identity. Wasm bindgen exports stateless JS wrapper functions which call into Wasm. Updating their internal
> Wasm instance binding to the new instance allows exposing the new Wasm instance without having to rebind the exported functions.

Author: ```suggestion
This function clears all internal state related to the Wasm VM, and updates all function bindings in place to reference the new WebAssembly instance.
```

---
title: Panic Recovery for Rust Workers
description: Making workers-rs deployments more reliable with panic recovery
date: 2025-09-19
---

import { WranglerConfig, Aside } from "~/components";

A long-standing problem we've had with [workers-rs](https://github.com/cloudflare/workers-rs) is that its Rust panics are non-recoverable, and can put the worker into an invalid state.

Working on the [wasm-bindgen](https://github.com/wasm-bindgen/wasm-bindgen) internals, we've been able to implement a solution for panic recovery to ensure more reliable deployments.

## Fixing Rust Panics with Wasm Bindgen

We use [wasm-bindgen](https://github.com/wasm-bindgen/wasm-bindgen) to build and deploy Rust workers, but Wasm Bindgen is not designed to support recoverable panics.

When a panic happens in Wasm Bindgen, the entire Wasm application is considered to be in an invalid state, and any further function calls could result in overflows or memory exceptions.

But Wasm is a virtual machine - if we can reset all the Wasm internal state back to its initial state then we can continue to take new requests.

> It's possible to add a panic handler in Rust to detect immediately when a panic happens:

Reviewer: I would be more direct here.

`You can now add panic handlers in Rust, like so:`

```rust
std::panic::set_hook(Box::new(move |panic_info| {
  hook_impl(panic_info);
}));
```

With some extra wiring we can then export a registration function to JavaScript to allow a custom panic hook to be attached:

```js

> import { setPanicHook } from "./index.js";

Reviewer: Where is this actually defined? - Not sure as a user where I go to do this.
